lib/compiler-llvm/src/translator/abi.rs

```diff
@@ -22,556 +21,1174 @@ use inkwell::{
-
-    let sig_returns_bitwidths = sig
-        .results()
-        .iter()
-        .map(|ty| match ty {
-            Type::I32 | Type::F32 => 32,
-            Type::I64 | Type::F64 => 64,
-            Type::V128 => 128,
-            Type::ExternRef => unimplemented!("externref in the llvm backend"),
-            Type::FuncRef => unimplemented!("funcref in the llvm backend"),
-        })
-        .collect::<Vec<i32>>();
-
-    Ok(match sig_returns_bitwidths.as_slice() {
-        [] => (
-            intrinsics
-                .void_ty
-                .fn_type(&param_types.collect::<Result<Vec<_>, _>>()?, false),
-            vec![],
-        ),
-        [_] => {
-            let single_value = sig.results()[0];
-            (
-                type_to_llvm(intrinsics, single_value)?
-                    .fn_type(&param_types.collect::<Result<Vec<_>, _>>()?, false),
-                vec![],
+/// We need to produce different LLVM IR for different platforms. (Contrary to
+/// popular knowledge LLVM IR is not intended to be portable in that way.) This
+/// trait deals with differences between function signatures on different
+/// targets.
+pub trait Abi {
+    /// Given a function definition, retrieve the parameter that is the vmctx pointer.
+    fn get_vmctx_ptr_param<'ctx>(&self, func_value: &FunctionValue<'ctx>) -> PointerValue<'ctx>;
+
+    /// Given a wasm function type, produce an llvm function declaration.
+    fn func_type_to_llvm<'ctx>(
+        &self,
+        context: &'ctx Context,
+        intrinsics: &Intrinsics<'ctx>,
+        sig: &FuncSig,
+    ) -> Result<(FunctionType<'ctx>, Vec<(Attribute, AttributeLoc)>), CompileError>;
+
+    /// Marshall wasm stack values into function parameters.
+    fn args_to_call<'ctx>(
+        &self,
+        alloca_builder: &Builder<'ctx>,
+        func_sig: &FuncSig,
+        ctx_ptr: PointerValue<'ctx>,
+        llvm_fn_ty: &FunctionType<'ctx>,
+        values: &[BasicValueEnum<'ctx>],
+    ) -> Vec<BasicValueEnum<'ctx>>;
+
+    /// Given a CallSite, extract the returned values and return them in a Vec.
+    fn rets_from_call<'ctx>(
+        &self,
+        builder: &Builder<'ctx>,
+        intrinsics: &Intrinsics<'ctx>,
+        call_site: CallSiteValue<'ctx>,
+        func_sig: &FuncSig,
+    ) -> Vec<BasicValueEnum<'ctx>>;
+
+    /// Whether the llvm equivalent of this wasm function has an `sret` attribute.
+    fn is_sret(&self, func_sig: &FuncSig) -> Result<bool, CompileError>;
+
+    /// Pack LLVM IR values representing individual wasm values into the return type for the function.
+    fn pack_values_for_register_return<'ctx>(
+        &self,
+        intrinsics: &Intrinsics<'ctx>,
+        builder: &Builder<'ctx>,
+        values: &[BasicValueEnum<'ctx>],
+        func_type: &FunctionType<'ctx>,
+    ) -> Result<BasicValueEnum<'ctx>, CompileError>;
+}
+
+/// Implementation of the [`Abi`] trait for the AMD64 SystemV ABI.
```

Comment:
Can we divide this into different files?
* `lib/compiler-llvm/src/translator/abi/x86_64_systemv.rs`
* `lib/compiler-llvm/src/translator/abi/aarch64_systemv.rs`

Reply:
Done.